icu4j/main/core/src/main/java/com/ibm/icu/util/IslamicCalendar.java

```diff
@@ -319,6 +320,294 @@ public IslamicCalendar(int year, int month, int date, int hour,
         this.set(Calendar.SECOND, second);
     }
 
+    // Private interface for different Islamic calendar algorithms.
+    private interface Algorithm {
+        /**
+         * Returns <code>true</code> if this object is using the fixed-cycle civil
+         * calendar, or <code>false</code> if using the religious, astronomical
+         * calendar.
+         */
+        public boolean isCivil();
+
+        /**
+         * Return the type the algorithm implement.
+         */
+        public CalculationType getType();
+
+        /**
+         * Return the epoch used by this algorithm.
+         */
+        public long epoch();
+
+        /**
+         * Return the day # on which the given year starts. Days are counted
+         * from the Hijri epoch, origin 0.
+         *
+         * @param year  The hijri year
+         */
+        public long yearStart(int year);
+
+        /**
+         * Return the day # on which the given month starts. Days are counted
+         * from the Hijri epoch, origin 0.
+         *
+         * @param year  The hijri year
+         * @param month  The hijri month, 0-based
+         */
+        public long monthStart(int year, int month);
+
+        /**
+         * Return the length (in days) of the given month.
+         *
+         * @param extendedYear  The hijri year
+         * @param month The hijri month, 0-based
+         */
+        public int monthLength(int extendedYear, int month);
```

Comment:
How's the "extended year" on this function different from the "year" on `monthStart()`?